mzLib/FlashLFQ/FlashLfqEngine.cs

```diff
@@ -855,36 +1080,95 @@ private void QuantifyMatchBetweenRunsPeaks(SpectraFileInfo idAcceptorFile)
                     continue;
                 }
 
-                List<ChromatographicPeak> peakHypotheses = mbrIdentifiedPeptide.Value.SelectMany(p => p.Value).OrderByDescending(p => p.MbrScore).ToList();
-
-                ChromatographicPeak best = peakHypotheses.First();
-
-                peakHypotheses.Remove(best);
-
-                if (peakHypotheses.Count > 0)
+                foreach (var peakHypothesisGroup in mbrIdentifiedPeptide.Value.SelectMany(kvp => kvp.Value).OrderByDescending(p => p.MbrScore).GroupBy(p => p.RandomRt))
                 {
-                    double start = best.IsotopicEnvelopes.Min(p => p.IndexedPeak.RetentionTime);
-                    double end = best.IsotopicEnvelopes.Max(p => p.IndexedPeak.RetentionTime);
+                    var peakHypotheses = peakHypothesisGroup.ToList();
+                    ChromatographicPeak best = peakHypotheses.First();
+                    peakHypotheses.Remove(best);
 
-                    List<ChromatographicPeak> peaksToRemoveFromHypotheses = new List<ChromatographicPeak>();
-                    foreach (ChromatographicPeak peak in peakHypotheses.Where(p => p.Apex.ChargeState != best.Apex.ChargeState))
+                    // Discard any peaks that are already associated with an ms/ms identified peptide
+                    while (best?.Apex?.IndexedPeak != null && msmsImsPeaks.TryGetValue(best.Apex.IndexedPeak.ZeroBasedMs1ScanIndex, out var peakList))
                     {
-                        if (peak.Apex.IndexedPeak.RetentionTime > start && peak.Apex.IndexedPeak.RetentionTime < end)
+                        if (peakList.Contains(best.Apex.IndexedPeak))
                         {
-                            best.MergeFeatureWith(peak, Integrate);
+                            if (!peakHypotheses.Any())
+                            {
+                                best = null;
+                                break;
+                            }
+                            best = peakHypotheses.First();
+                            peakHypotheses.Remove(best);
+                        }
```

Comment:
were these sections not tested previously ?

Reply:
Not directly, no. A lot of this coed existed previously but was refactored to make it more modular